onaio/onadata
# Flexible mqtt topics

### Changes / Features implemented
Supercedes https://github.com/onaio/onadata/pull/1793
1. Remove send message from post save signal. 
    - For forms, this is called each time there is an update to the form yet the notification is only interested in changes in the form xls file itself.
    - For submissions, this isn't called when there is a bulk update of the submissions.
2. Forms: Only send a notification when the form xls has been updated
    Submissions: Send notification both upon creation and edit
3. Send notification upon bulk deletion during a csv import.
4. Send notification on bulk submission review
5. Change topics to allow more flexibility. The topics will be of the form:
    `/onadata/organization/{org_id}/project/{proj_id}/xform/{xform_id}/#`

### To test
1. MQTT server
- ssh to mqtt server
- listen to any of the channels below
- perform the relevant action (delete, create, edit ...)
- a message payload should be posted on the channel

2. API
- perform the relevant action (delete, create, edit ...)
- a message payload should be included in the relevant API json response 

|channel | url| action|
|---------|--|--------|
|`/onadata/organization/[org_username]/project/[proj_id]/xform/[form_id]/submission/deleted`| `/api/v1/messaging.json?target_type=xform&target_id={form_id}&verb=submission_deleted`|submission deletion|
|`/onadata/organization/[org_username]/project/[proj_id]/xform/[form_id]/submission/created `|`/api/v1/messaging.json?target_type=xform&target_id={form_id}&verb=submission_created`| submission creation |
|`/onadata/organization/[org_username]/project/[proj_id]/xform/[form_id]/submission/edited `| `/api/v1/messaging.json?target_type=xform&target_id={form_id}&verb=submission_edited`|submission edit |
|`/onadata/organization/[org_username]/project/[proj_id]/xform/[form_id]/submission/reviewed `| `/api/v1/messaging.json?target_type=xform&target_id={form_id}&verb=submission_review`|submission_review|
|`/onadata/organization/[org_username]/project/[proj_id]/xform/[form_id]/form/updated `| `/api/v1/messaging.json?target_type=xform&target_id={form_id}&verb=form_updated`|form updated|

Example payload:
```json
{
    "id": 113, 
    "verb": 
    "submission_deleted", 
    "message": {"id": ["35838"]}, 
    "user": "linc", 
    "timestamp": "2020-06-10T11:10:45.346353+00:00"
}
```
### Steps taken to verify this change does what is intended
Written tests
### Side effects of implementing this change

Closes #1846
